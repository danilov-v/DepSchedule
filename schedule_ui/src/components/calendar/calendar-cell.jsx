import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import "./calendar.scss";

export function CalendarCell({ text, fluid }) {
  return (
    <div
      className={classnames(
        "calendar-cell d-flex align-items-center justify-content-center",
        {
          fluid,
        }
      )}
    >
      {text}
    </div>
  );
}

CalendarCell.propTypes = {
  text: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.node,
  ]),
  fluid: PropTypes.bool,
};

CalendarCell.defaultProps = {
  text: null,
  fluid: false,
};
