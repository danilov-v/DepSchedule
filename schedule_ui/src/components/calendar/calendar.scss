.calendar {
  position: sticky;
  top: 60px;
  background: #f2f3f2;
  font-size: 1.1rem;
  font-weight: 550;
  word-break: break-all;
  z-index: 1;

  &-cell {
    width: 35px;
    height: 35px;
    border: 1px solid black;

    &.fluid {
      width: 100%;
    }
  }
}
