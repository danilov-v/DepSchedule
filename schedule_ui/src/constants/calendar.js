export const MONTH_TRANSLATIONS = {
  January: "Январь",
  February: "Февраль",
  March: "Март",
  April: "Апрель",
  May: "Май",
  June: "Июнь",
  July: "Июль",
  August: "Август",
  September: "Сентябрь",
  October: "Октябрь",
  November: "Ноябрь",
  December: "Декабрь",
};

export const EVENT_TYPE_COLORS = [
  "#D7FFAB",
  "#67ecc0",
  "#ffb2b6",
  "#ff5555",
  "#b6ddfb",
  "#fff292",
  "#b9d7d8",
  "#33b7ec",
  "#d0ccff",
  "#dcb684",
];

export const FOUR_MONTH = 122;
