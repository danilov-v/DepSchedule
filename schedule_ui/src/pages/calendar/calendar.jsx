import React from "react";
import { Home } from "components/home/home";

import { useCalendars } from "helpers/hooks/apiEffects";

// TODO: implement logic when BE will work
const getActiveCalendar = calendars => calendars && calendars[0];

export function Calendar() {
  const [calendars] = useCalendars();
  const activeCalendar = getActiveCalendar(calendars);

  if (!activeCalendar) return null;

  return <Home activeCalendar={activeCalendar} />;
}
