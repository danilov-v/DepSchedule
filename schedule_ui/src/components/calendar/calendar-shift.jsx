import React, { Fragment } from "react";
import { isWithinInterval, differenceInDays } from "date-fns";
import { Row, Col } from "reactstrap";
import { useSelector } from "react-redux";
import { getFormattedPeriodsSelector } from "redux/selectors/periods";
import { CalendarCell } from "./calendar-cell";

const getPeriod = (day, periods) =>
  periods.find(period =>
    isWithinInterval(day, { start: period.startDate, end: period.endDate })
  );

const getShift = (day, period) => differenceInDays(day, period.startDate);

export function CalendarShift({ range }) {
  const periods = useSelector(getFormattedPeriodsSelector);

  return (
    <Row className="calendar calendar-shift flex-nowrap" noGutters>
      {range.map(month => (
        <Fragment key={month.name}>
          {month.days.map(day => {
            const period = getPeriod(day, periods);
            const shift = period && getShift(day, period).toString();

            return (
              <Col key={day.getTime()}>
                <CalendarCell
                  text={shift ? `M${shift}` : ""}
                  className={
                    shift ? "calendar-cell-shift" : "calendar-cell-empty"
                  }
                />
              </Col>
            );
          })}
        </Fragment>
      ))}
    </Row>
  );
}
