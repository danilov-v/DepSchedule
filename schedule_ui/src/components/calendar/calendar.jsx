import React, { Fragment } from "react";
import PropTypes from "prop-types";
import { Row, Col } from "reactstrap";
import { CalendarCell } from "./calendar-cell";
import { UnitEventRow } from "../unit-event-row/unit-event-row";

import "./calendar.scss";

export function Calendar({ range, unitGroups, showMonth }) {
  return (
    <Fragment>
      <Row className="calendar" noGutters>
        {range.map(month => (
          <Col key={month.name}>
            {showMonth && (
              <Row className="calendar-title" noGutters>
                <Col>
                  <CalendarCell text={month.name} fluid />
                </Col>
              </Row>
            )}
            <Row noGutters>
              <Col className="d-flex">
                {month.days.map(day => (
                  <CalendarCell
                    key={day.getTime()}
                    text={showMonth ? day.getDate() : null}
                  />
                ))}
              </Col>
            </Row>
          </Col>
        ))}
      </Row>
      {unitGroups.map((group, i) => (
        <UnitEventRow key={i} range={range} unitGroup={group} />
      ))}
    </Fragment>
  );
}

Calendar.propTypes = {
  range: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      days: PropTypes.arrayOf(PropTypes.object),
    })
  ),
  showMonth: PropTypes.bool,
};

Calendar.defaultProps = {
  range: [],
  showMonth: false,
};
