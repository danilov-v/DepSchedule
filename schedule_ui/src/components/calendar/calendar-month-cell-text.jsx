import React from "react";
import { CELL_WIDTH, CELL_WIDTH_BORDER } from "constants/calendar";

export function CalendarMonthCellText({ text, length }) {
  return (
    <p
      className="m-0 calendar-month-cell-text"
      style={{ maxWidth: length * (CELL_WIDTH - CELL_WIDTH_BORDER) }}
    >
      {text}
    </p>
  );
}
