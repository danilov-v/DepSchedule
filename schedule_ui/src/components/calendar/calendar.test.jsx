import React from "react";
import { shallow } from "enzyme";
import { Calendar } from "./calendar";

const data = [{ name: "November", days: [new Date(2019, 7, 7)] }];

describe("calendar.jsx", () => {
  describe("<Calendar />", () => {
    it("should render and match snapshot", () => {
      const component = shallow(<Calendar range={data} />);

      expect(component).toMatchSnapshot();
    });
  });
});
