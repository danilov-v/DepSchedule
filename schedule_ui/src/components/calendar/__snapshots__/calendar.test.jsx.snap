// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calendar.jsx <Calendar /> should render and match snapshot 1`] = `
<Fragment>
  <Row
    className="calendar"
    noGutters={true}
    tag="div"
  >
    <Col
      key="November"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Row
        noGutters={true}
        tag="div"
      >
        <Col
          className="d-flex"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <CalendarCell
            fluid={false}
            key="1565125200000"
            text={null}
          />
        </Col>
      </Row>
    </Col>
  </Row>
  <UnitEventRow
    key="0"
    range={
      Array [
        Object {
          "days": Array [
            2019-08-06T21:00:00.000Z,
          ],
          "name": "November",
        },
      ]
    }
    unitGroup={
      Array [
        Object {
          "parentId": 201,
          "title": "Система управления 1.1.1",
          "unitId": 251,
          "unitLevel": 3,
        },
        Object {
          "parentId": 201,
          "title": "Система управления 1.1.2",
          "unitId": 252,
          "unitLevel": 3,
        },
      ]
    }
  />
  <UnitEventRow
    key="1"
    range={
      Array [
        Object {
          "days": Array [
            2019-08-06T21:00:00.000Z,
          ],
          "name": "November",
        },
      ]
    }
    unitGroup={
      Array [
        Object {
          "parentId": 301,
          "title": "Система управления 2.1.1",
          "unitId": 2522,
          "unitLevel": 3,
        },
        Object {
          "parentId": 301,
          "title": "Система управления 2.1.2",
          "unitId": 2533,
          "unitLevel": 3,
        },
      ]
    }
  />
</Fragment>
`;
